src/tir/transforms/lower_sparse_tir.cc

```diff
@@ -260,29 +322,146 @@ class IndexTransformer : public StmtExprMutator {
+    Optional<Stmt> init =
+        sp_block->init.defined() ? VisitStmt(sp_block->init.value()) : Optional<Stmt>(NullOpt);
+    Stmt body = VisitStmt(sp_block->body);
+
+    // Step 2. Create the new outer loop vars.
+    Array<Var> loop_vars;
+    std::unordered_map<const VarNode*, PrimExpr> var_map;
+    loop_vars.reserve(n_iter);
+    var_map.reserve(n_iter);
+    for (const SpIterVar& sp_iter : sp_block->sp_iter_vars) {
+      Var loop_var("v_" + sp_iter->var->name_hint);
+      loop_vars.push_back(loop_var);
+      var_map[sp_iter->var.get()] = loop_var;
+    }
+
+    // Step 3. Create block iters and iter bindings.
+    Array<IterVar> block_iters;
+    Array<PrimExpr> iter_bindings;
+    block_iters.reserve(n_iter);
+    iter_bindings.reserve(n_iter);
+    for (int i = 0; i < n_iter; ++i) {
+      block_iters.push_back(SpIterVar2IterVar(sp_block->sp_iter_vars[i], var_map));
+      iter_bindings.push_back(loop_vars[i]);
+    }
+
+    // Step 4. Generate the read-region and write-retion of the block.
+    Array<BufferRegion> reads{nullptr};
+    Array<BufferRegion> writes{nullptr};
+    GenerateReadWriteRegions(sp_block, &reads, &writes);
+
+    // Step 5. Create the block and block-realize
+    Block block(block_iters, std::move(reads), std::move(writes), sp_block->name, std::move(body),
+                std::move(init));
+    BlockRealize block_realize(std::move(iter_bindings), const_true(), std::move(block));
+
+    // Step 6. Create outer loops and the block binding.
+    Stmt loop = GenerateLoops(std::move(block_realize), block_iters, loop_vars);
+
+    return loop;
+  }
+
+  IterVar SpIterVar2IterVar(const SpIterVar& sp_iter,
+                            const std::unordered_map<const VarNode*, PrimExpr>& var_map) {
+    PrimExpr extent{nullptr};
+
+    SpIterKind kind = sp_iter->kind;
+    if (kind == SpIterKind::kDenseFixed || kind == SpIterKind::kSparseFixed) {
+      extent = sp_iter->max_extent;
+    } else {
+      SparseBuffer iterated_buffer{nullptr};
+      Array<PrimExpr> dependent_iters{nullptr};
+      collector_.GetIteratedBufferAndDependentIters(sp_iter, &iterated_buffer, &dependent_iters);
+      PrimExpr lowered_indices = LowerIndices(std::move(iterated_buffer), dependent_iters);
+
+      Buffer indptr{kind == SpIterKind::kDenseVariable
+                        ? Downcast<DenseVariableAxis>(sp_iter->axis)->indptr
+                        : Downcast<SparseVariableAxis>(sp_iter->axis)->indptr};
+      PrimExpr l = BufferLoad(indptr, {lowered_indices});
+      PrimExpr r = BufferLoad(indptr, {add(lowered_indices, 1)});
+      extent = sub(r, l);
+    }
+
+    // Substitute the iteration vars in the expression with the loop vars.
+    return IterVar(Range::FromMinExtent(0, Substitute(std::move(extent), var_map)), sp_iter->var,
+                   sp_iter->is_reduction ? kCommReduce : kDataPar);
+  }
+
+  void GenerateReadWriteRegions(const SparseBlockNode* sp_block, Array<BufferRegion>* reads,
```

Comment:
We might refine this part in the future.

Reply:
Sure.